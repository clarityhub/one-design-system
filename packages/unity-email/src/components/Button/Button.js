import React from 'react';
import {
	MjmlButton,
} from 'mjml-react';
import { node } from 'prop-types';
import colors from '@clarityhub/unity-core/dist/theme/colors';

import { variants } from '../../theme/fonts';

const { letterSpacing, ...validButtonOptions } = variants.button;

const buttonBase = Object.assign(
	{},
	validButtonOptions,
	{
		fontSize: '16px',
		borderRadius: '4px', // TODO merge with borders.borderRadius
		paddingTop: '8px',
		paddingBottom: '8px',
		paddingLeft: '20px',
		paddingRight: '20px',
	}
);

const determineColors = (type) => {
	switch (type) {
	case 'primary':
		return {
			color: colors.white,
			backgroundColor: colors.primary,
		};
	case 'danger':
		return {
			color: colors.white,
			backgroundColor: colors.danger,
		};
	case 'default':
	default:
		return {
			color: colors.primary,
			backgroundColor: 'rgba(255, 255, 255, 0)',
		};
	}
};

const Button = ({ children, center, type, ...rest }) => (
	<MjmlButton {...buttonBase} {...determineColors(type)} align={center ? 'center' : 'left'} textAlign={center ? 'left': 'center'} {...rest}>
		{children}
	</MjmlButton>
);

Button.propType = {
	children: node,
};

export default Button;
