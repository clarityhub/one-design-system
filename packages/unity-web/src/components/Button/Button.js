import React from 'react';
import cx from 'classnames';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import { bool, func, oneOf, oneOfType, node, string } from 'prop-types';
import { variants } from '@clarityhub/unity-core/lib/typography';
import borders from '@clarityhub/unity-core/lib/borders';
import colors from '@clarityhub/unity-core/lib/colors';
import { lighten, darken, desaturate } from '@clarityhub/unity-core/lib/utilities/color';

import Loading from '../svgs/Loading';

const Button = styled.button(
	css`
        overflow: visible;
        display: inline-block;

        ${variants.button.string}

        box-shadow: ${colors.shadow.default};
        cursor: pointer;
        text-align: center;
        white-space: nowrap;
        vertical-align: middle;
        user-select: none;
        border: 1px solid transparent;
        padding: .5rem 1.25rem;
        font-size: 1rem;
        line-height: 1.5;
        border-radius: ${borders.borderRadius.rem}rem;
        position: relative;
        text-decoration: none;
        text-transform: uppercase;
        transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;

        &:focus,
        &:hover {
            text-decoration: none;
        }

        &:focus,
        &.focused {
            outline: 0;
            box-shadow: 0 0 0 .2rem ${colors.focused.default};
        }

        &:disabled,
        &.disabled {
            cursor: not-allowed;
            pointer-events: none;
        }

        &:not(:disabled):not(.disabled) {
            cursor: pointer;
        }

        svg {
            height: 1rem;
            width: 1rem;
        }

        --text: ${colors.primary.default};
        --text-active: ${colors.black.default};
        --bg: transparent;
        --bg-active: transparent;
        --bg-disabled: transparent;

        &.-default,
        &.-primary,
        &.-danger,
        &.-white {
            color: var(--text);
            background-color: var(--bg);
            border-color: var(--bg);

            &.-outline {
                border-color: var(--outline, var(--bg));
                color: var(--outline-text, var(--bg));
            }

            &.-text {
                color: var(--text-text, var(--bg));
            }

            &:hover,
            &.active {
                background-color: var(--bg-active);
                border-color: var(--bg-active);
                color: var(--text-active);

                &.-outline {
                    background-color: var(--bg) !important;
                    border: 1px solid var(--bg);
                    color: var(--outline-text-active, var(--text));
                }

                &.-text {
                    color: var(--text-text-active, var(--bg-active));
                }
            }

            &:disabled,
            &.disabled {
                background-color: var(--bg-disabled);
                border-color: var(--bg-disabled);
                color: var(--text);

                &.-outline {
                    border: 1px solid var(--outline-disabled, var(--outline-text-disabled, var(--bg-disabled)));
                    color: var(--outline-text-disabled, var(--bg-disabled));
                }

                &.-text {
                    color: var(--text-text-disabled, var(--bg-disabled));
                }
            }
        }

        &.-primary {
            --text: ${colors.white.default};
            --text-active: ${colors.white.default};
            --outline-disabled: ${colors.gray.default};
            --bg: ${colors.primary.default};
            --bg-active: ${desaturate(darken(colors.primary.default))};
            --bg-disabled: ${desaturate(lighten(colors.primary.default))};
        }

        &.-danger {
            --text: ${colors.white.default};
            --text-active: ${colors.white.default};
            --outline-disabled: ${colors.gray.default};
            --bg: ${colors.danger.default};
            --bg-active: ${desaturate(darken(colors.danger.default))};
            --bg-disabled: ${lighten(colors.danger.default, 0.3)};
        }

        &.-default {
            --text: ${colors.black.default};
            --text-active: ${colors.black.default};
            --outline: ${colors.black.default};
            --outline-text: ${colors.black.default};
            --outline-text-active: ${colors.black.default};
            --outline-text-disabled: ${colors.gray.default};
            --text-text: ${colors.black.default};
            --text-text-active: ${colors.black.default};
            --text-text-disabled: ${colors.gray.default};
            --bg: ${colors.white.default};
            --bg-active: ${colors.dove.default};
            --bg-disabled: ${colors.muted.default};
        }

        &.-white {
            --text: ${colors.primary.default};
            --text-active: ${colors.black.default};
            --outline: ${colors.white.default};
            --outline-text: ${colors.white.default};
            --outline-text-active: ${colors.primary.default};
            --outline-text-disabled: ${colors.gray.default};
            --text-text: ${colors.white.default};
            --text-text-active: ${colors.white.default};
            --text-text-disabled: ${colors.gray.default};
            --bg: ${colors.white.default};
            --bg-active: ${colors.dove.default};
            --bg-disabled: ${colors.muted.default};
        }

        &.-text {
            background-color: transparent !important;
            border: 1px solid transparent;
            box-shadow: none;

            &:hover,
            &.active {
                border-color: transparent;
            }

            &.disabled,
            &:disabled {
                border: 1px solid transparent;
            }
        }

        &.-outline {
            background-color: transparent !important;
            border: 1px solid inherit;
            box-shadow: none;
        }

        &.disabled,
        &:disabled {
            cursor: not-allowed;
        }
    `,
	({ size }) => {
		switch (size) {
		case 'small':
			return css`
                font-size: 0.9rem;
                padding: .2rem 1rem;
            `;
		case 'large':
			return css`
                font-size: 1.2rem;
                padding: 0.8rem 1.4rem;
            `;
		case 'default':
		default:
			return '';
		}
	},

	({ block }) => block && css`
        display: block;
        width: 100%;
    `,

	({ loading }) => loading && css`
        color: transparent !important;
        cursor: wait !important;

        &:hover {
            color: transparent !important;
        }
    `,
);

const IconWrapper = styled.span`
    font-size: inherit;
    margin-right: 0.3rem;
    vertical-align: middle;

    svg {
        height: 1rem;
        width: 1rem;

        ${({ size }) => {
		switch (size) {
		case 'small':
			return css`
                height: 0.9rem;
                width: 0.9rem;
            `;
		case 'large':
			return css`
                height: 1.2rem;
                width: 1.2rem;
            `;
		case 'default':
		default:
			return '';
		}
	}}
    }
`;

const LoadingWrapper = styled.div`
    position: absolute;
    left: 50%;
    transform: translate(-50%, -50%);
    top: 50%;
    align-content: center;
    justify-content: center;
    display: flex;

    svg {
        height: 1.4rem;
        width: 1.4rem;

        ${({ size }) => {
		switch (size) {
		case 'small':
			return css`
                height: 1.2rem;
                width: 1.2rem;
            `;
		case 'large':
			return css`
                height: 2.4rem;
                width: 2.4rem;
            `;
		case 'default':
		default:
			return '';
		}
	}}

    path {
        ${({ type, outline, text }) => {
		switch (type) {
		case 'primary':
			return css`
                fill: ${colors.white.default};

                ${outline && css`
                    fill: ${colors.primary.default};
                `}

                ${text && css`
                    fill: ${colors.primary.default};
                `}

                &:hover,
                &.active {
                    ${outline && css`
                        fill: ${colors.white.default};
                    `}

                    ${text && css`
                        fill: ${desaturate(darken(colors.primary.default))};
                    `}
                }

                &:disabled,
                &.disabled {
                    ${outline && css`
                        fill: ${desaturate(lighten(colors.primary.default))};
                    `}

                    ${text && css`
                        fill: ${desaturate(lighten(colors.primary.default))};
                    `}
                }
            `;
		case 'danger':
			return css`
                fill: ${colors.white.default};

                ${outline && css`
                    fill: ${colors.danger.default};
                `}

                ${text && css`
                    fill: ${colors.danger.default};
                `}

                &:hover,
                &.active {
                    ${outline && css`
                        fill: ${colors.white.default};
                    `}

                    ${text && css`
                        fill: ${desaturate(darken(colors.danger.default))};
                    `}
                }

                &:disabled,
                &.disabled {
                    ${outline && css`
                        fill: ${desaturate(lighten(colors.danger.default, 0.3))};
                    `}

                    ${text && css`
                        fill: ${desaturate(lighten(colors.danger.default, 0.3))};
                    `}
                }
            `;
		case 'default':
		default:
			return css`
                fill: ${colors.black.default};

                ${outline && css`
                    fill: ${colors.black.default};
                `}

                ${text && css`
                    fill: ${colors.black.default};
                `}

                &:hover,
                &.active {
                    ${outline && css`
                        fill: ${colors.black.default};
                    `}

                    ${text && css`
                        fill: ${desaturate(darken(colors.black.default))};
                    `}
                }

                &:disabled,
                &.disabled {
                    fill: ${darken(colors.gray.default, 0.1)};

                    ${outline && css`
                        fill: ${lighten(colors.gray.default, 0.01)};
                    `}

                    ${text && css`
                        fill: ${desaturate(lighten(colors.black.default))};
                    `}
                }
            `;
		}}
}
        }
    }
`;

const renderIcon = (Icon) => {
	if (typeof Icon === 'function') {
		return <Icon />;
	}

	return Icon;
};

const ButtonWrapper = ({
	className,
	type,
	outline,
	text,
	size,
	icon,
	children, 
	component,
	loading,
	buttonType = 'button',
	...rest
}) => {
	let Component = Button;

	if (component) {
		Component = Button.withComponent(component);
	}

	return (
		<Component
			className={cx(className, {
				[`-${type}`]: type,
				'-outline': outline,
				'-text': text,
			})}
			outline={outline}
			text={text}
			size={size}
			loading={loading}
			type={buttonType}
			{...rest}
		>
			{
				icon && (
					<IconWrapper size={size}>{renderIcon(icon)}</IconWrapper>
				)
			}
			{
				loading && (
					<LoadingWrapper
						type={type}
						outline={outline}
						text={text}
						size={size}
						{...rest}
					>
						<Loading />
					</LoadingWrapper>
				)
			}
			{children}
		</Component>
	);
};

ButtonWrapper.propTypes = {
	block: bool,
	buttonType: string,
	component: oneOfType([string, func]),
	icon: oneOfType([node, func]),
	loading: bool,
	outline: bool,
	size: oneOf(['small', 'default', 'large']),
	text: bool,
	type: oneOf(['default', 'white', 'primary', 'danger', 'success']),
};

ButtonWrapper.defaultProps = {
	type: 'default',
};

export { Button as EmotionButton };
export default ButtonWrapper;
