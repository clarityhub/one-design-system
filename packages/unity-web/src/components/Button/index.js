export { default, EmotionButton } from './Button';