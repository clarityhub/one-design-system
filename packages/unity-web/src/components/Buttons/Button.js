import React from 'react';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import { bool, func, oneOf, oneOfType, node, string } from 'prop-types';
import { variants } from '@clarityhub/unity-core/lib/typography';
import borders from '@clarityhub/unity-core/lib/borders';
import colors from '@clarityhub/unity-core/lib/colors';
import { lighten, darken, desaturate } from '@clarityhub/unity-core/lib/utilities/color';

import Loading from '../svgs/Loading';

const Button = styled.button(
	css`
        overflow: visible;
        display: inline-block;

        ${variants.button.string}

        box-shadow: ${colors.shadow.default};
        text-align: center;
        white-space: nowrap;
        vertical-align: middle;
        user-select: none;
        border: 1px solid transparent;
        padding: .5rem 1.25rem;
        font-size: 1rem;
        line-height: 1.5;
        border-radius: ${borders.borderRadius.rem}rem;
        position: relative;
        text-decoration: none;
        text-transform: uppercase;
        transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;

        &:focus,
        &:hover {
            text-decoration: none;
        }

        &:focus,
        &.focused {
            outline: 0;
            box-shadow: 0 0 0 .2rem ${colors.highlight.default};
        }

        &:disabled,
        &.disabled {
            cursor: not-allowed;
            pointer-events: none;
        }

        &:not(:disabled):not(.disabled) {
            cursor: pointer;
        }
    `,
	({ type, outline, text }) => {
		switch (type) {
		case 'primary':
			return css`
                    color: ${colors.white.default};
                    background-color: ${colors.primary.default};
                    border-color: ${colors.primary.default};

                    ${outline && `
                        background-color: transparent;
                        border: 1px solid inherit;
                        color: ${colors.primary.default};
                    `}

                    ${text && `
                        background-color: transparent;
                        border: 1px solid transparent;
                        color: ${colors.primary.default};
                    `}

                    &:hover,
                    &.active {
                        background-color: ${desaturate(darken(colors.primary.default))};
                        border-color: ${desaturate(darken(colors.primary.default))};

                        ${outline && `
                            background-color: ${colors.primary.default};
                            border: 1px solid ${colors.primary.default};
                            color: ${colors.white.default};
                        `}

                        ${text && `
                            background-color: transparent;
                            border: 1px solid transparent;
                            color: ${desaturate(darken(colors.primary.default))};
                        `}
                    }

                    &:disabled,
                    &.disabled {
                        background-color: ${desaturate(lighten(colors.primary.default))};
                        border-color: ${desaturate(lighten(colors.primary.default))};

                        ${outline && `
                            background-color: transparent;
                            border: 1px solid ${colors.primary.default};
                            color: ${desaturate(lighten(colors.primary.default))};
                        `}

                        ${text && `
                            background-color: transparent;
                            border: 1px solid transparent;
                            color: ${desaturate(lighten(colors.primary.default))};
                        `}
                    }
                `;
		case 'danger':
			return css`
                    color: ${colors.white.default};
                    background-color: ${colors.danger.default};
                    border-color: ${colors.danger.default};

                    ${outline && `
                        background-color: transparent;
                        border: 1px solid inherit;
                        color: ${colors.danger.default};
                    `}

                    ${text && `
                        background-color: transparent;
                        border: 1px solid transparent;
                        color: ${colors.danger.default};
                    `}

                    &:hover,
                    &.active {
                        background-color: ${desaturate(darken(colors.danger.default))};
                        border-color: ${desaturate(darken(colors.danger.default))};

                        ${outline && `
                            background-color: ${colors.danger.default};
                            border: 1px solid ${colors.danger.default};
                            color: ${colors.white.default};
                        `}

                        ${text && `
                            background-color: transparent;
                            border: 1px solid transparent;
                            color: ${desaturate(darken(colors.danger.default))};
                        `}
                    }

                    &:disabled,
                    &.disabled {
                        background-color: ${lighten(colors.danger.default, 0.3)};
                        border-color: ${lighten(colors.danger.default, 0.3)};

                        ${outline && `
                            background-color: transparent;
                            border: 1px solid ${colors.danger.default};
                            color: ${desaturate(lighten(colors.danger.default, 0.3))};
                        `}

                        ${text && `
                            background-color: transparent;
                            border: 1px solid transparent;
                            color: ${desaturate(lighten(colors.danger.default, 0.3))};
                        `}
                    }
                `;
		case 'white':
			return css`
                background-color: ${colors.white.default};
                border-color: transparent;
                color: ${colors.primary.default};

                ${outline && `
                    background-color: transparent;
                    border: 1px solid inherit;
                    color: ${colors.white.default};
                `}

                ${text && `
                    background-color: transparent;
                    border: 1px solid transparent;
                    color: ${colors.white.default};
                `}

                &:hover,
                &.active {
                    background-color: ${colors.dove.default};
                    border-color: transparent;
                    color: ${colors.dark.default};

                    ${outline && `
                        background-color: ${colors.dove.default};
                        border: 1px solid ${colors.dove.default};
                        color: ${colors.black.default};
                    `}

                    ${text && `
                        background-color: transparent;
                        border: 1px solid transparent;
                        color: ${colors.dove.default};
                    `}
                }

                &:disabled,
                &.disabled {
                    color: ${darken(colors.gray.default, 0.1)};
                    background-color: ${lighten(colors.muted.default, 0.01)};
                    border-color: transparent;

                    ${outline && `
                        background-color: transparent;
                        border: 1px solid ${colors.gray.default};
                        color: ${lighten(colors.gray.default, 0.01)};
                    `}

                    ${text && `
                        background-color: transparent;
                        border: 1px solid transparent;
                        color: ${desaturate(lighten(colors.black.default))};
                    `}
                }
            `;
		case 'default':
		default:
			return css`
                    background-color: rgba(255, 255, 255, 0);
                    border-color: transparent;
                    color: ${colors.primary.default};

                    ${outline && `
                        background-color: transparent;
                        border: 1px solid inherit;
                        color: ${colors.black.default};
                    `}

                    ${text && `
                        background-color: transparent;
                        border: 1px solid transparent;
                        color: ${colors.black.default};
                    `}

                    &:hover,
                    &.active {
                        background-color: ${colors.dove.default};
                        border-color: transparent;
                        color: ${colors.dark.default};

                        ${outline && `
                            background-color: ${colors.muted.default};
                            border: 1px solid ${colors.muted.default};
                            color: ${colors.black.default};
                        `}

                        ${text && `
                            background-color: transparent;
                            border: 1px solid transparent;
                            color: ${desaturate(lighten(colors.black.default, 0.5))};
                        `}
                    }

                    &:disabled,
                    &.disabled {
                        color: ${darken(colors.gray.default, 0.1)};
                        background-color: ${lighten(colors.muted.default, 0.01)};
                        border-color: transparent;

                        ${outline && `
                            background-color: transparent;
                            border: 1px solid ${colors.gray.default};
                            color: ${lighten(colors.gray.default, 0.01)};
                        `}

                        ${text && `
                            background-color: transparent;
                            border: 1px solid transparent;
                            color: ${desaturate(lighten(colors.black.default))};
                        `}
                    }
                `;
		}
	},

	({ outline }) => outline && css`
        background-color: transparent;
        border: 1px solid inherit;
        box-shadow: none;
    `,

	({ text }) => text && css`
        box-shadow: none;
    `,

	({ size }) => {
		switch (size) {
		case 'small':
			return css`
                    font-size: 0.9rem;
                    padding: .2rem 1rem;
                `;
		case 'large':
			return css`
                    font-size: 1.2rem;
                    padding: 0.8rem 1.4rem;
                `;
		case 'default':
		default:
			return '';
		}
	},

	({ block }) => block && css`
        display: block;
        width: 100%;
    `,

	({ loading }) => loading && css`
        color: transparent !important;

        &:hover {
            color: transparent !important;
        }
    `,
);

const IconWrapper = styled.span`
    font-size: inherit;
    margin-right: 0.3rem;
    vertical-align: middle;

    svg {
        height: 1rem;
        width: 1rem;

        ${({ size }) => {
		switch (size) {
		case 'small':
			return css`
                        height: 0.9rem;
                        width: 0.9rem;
                    `;
		case 'large':
			return css`
                        height: 1.2rem;
                        width: 1.2rem;
                    `;
		case 'default':
		default:
			return '';
		}
	}}
    }
`;

const LoadingWrapper = styled.div`
    position: absolute;
    left: 50%;
    transform: translate(-50%, -50%);
    top: 50%;
    align-content: center;
    justify-content: center;
    display: flex;

    svg {
        height: 1.4rem;
        width: 1.4rem;

        ${({ size }) => {
		switch (size) {
		case 'small':
			return css`
                            height: 1.2rem;
                            width: 1.2rem;
                        `;
		case 'large':
			return css`
                            height: 2.4rem;
                            width: 2.4rem;
                        `;
		case 'default':
		default:
			return '';
		}
	}}

        path {
            ${({ type, outline, text }) => {
		switch (type) {
		case 'primary':
			return css`
                            fill: ${colors.white.default};

                            ${outline && css`
                                fill: ${colors.primary.default};
                            `}

                            ${text && css`
                                fill: ${colors.primary.default};
                            `}

                            &:hover,
                            &.active {
                                ${outline && css`
                                    fill: ${colors.white.default};
                                `}

                                ${text && css`
                                    fill: ${desaturate(darken(colors.primary.default))};
                                `}
                            }

                            &:disabled,
                            &.disabled {
                                ${outline && css`
                                    fill: ${desaturate(lighten(colors.primary.default))};
                                `}

                                ${text && css`
                                    fill: ${desaturate(lighten(colors.primary.default))};
                                `}
                            }
                        `;
		case 'danger':
			return css`
                            fill: ${colors.white.default};

                            ${outline && css`
                                fill: ${colors.danger.default};
                            `}

                            ${text && css`
                                fill: ${colors.danger.default};
                            `}

                            &:hover,
                            &.active {
                                ${outline && css`
                                    fill: ${colors.white.default};
                                `}

                                ${text && css`
                                    fill: ${desaturate(darken(colors.danger.default))};
                                `}
                            }

                            &:disabled,
                            &.disabled {
                                ${outline && css`
                                    fill: ${desaturate(lighten(colors.danger.default, 0.3))};
                                `}

                                ${text && css`
                                    fill: ${desaturate(lighten(colors.danger.default, 0.3))};
                                `}
                            }
                        `;
		case 'default':
		default:
			return css`
                            fill: ${colors.black.default};

                            ${outline && css`
                                fill: ${colors.black.default};
                            `}

                            ${text && css`
                                fill: ${colors.black.default};
                            `}

                            &:hover,
                            &.active {
                                ${outline && css`
                                    fill: ${colors.black.default};
                                `}

                                ${text && css`
                                    fill: ${desaturate(darken(colors.black.default))};
                                `}
                            }

                            &:disabled,
                            &.disabled {
                                fill: ${darken(colors.gray.default, 0.1)};

                                ${outline && css`
                                    fill: ${lighten(colors.gray.default, 0.01)};
                                `}

                                ${text && css`
                                    fill: ${desaturate(lighten(colors.black.default))};
                                `}
                            }
                        `;
		}}
}
        }
    }
`;

const renderIcon = (Icon) => {
	if (typeof Icon === 'function') {
		return <Icon />;
	}

	return Icon;
};

const ButtonWrapper = ({ icon, children, component, size, loading, buttonType = 'button', ...rest }) => {
	let Component = Button;

	if (component) {
		Component = Button.withComponent(component);
	}

	return (
		<Component size={size} loading={loading} {...rest}>
			{
				icon && (
					<IconWrapper size={size}>{renderIcon(icon)}</IconWrapper>
				)
			}
			{
				loading && (
					<LoadingWrapper size={size} {...rest}><Loading /></LoadingWrapper>
				)
			}
			{children}
		</Component>
	);
};

ButtonWrapper.propTypes = {
	block: bool,
	buttonType: string,
	component: oneOfType([string, func]),
	icon: oneOfType([node, func]),
	loading: bool,
	outline: bool,
	size: oneOf(['small', 'default', 'large']),
	text: bool,
	type: oneOf(['default', 'white', 'primary', 'danger', 'success']),
};

export { Button as EmotionButton };
export default ButtonWrapper;
