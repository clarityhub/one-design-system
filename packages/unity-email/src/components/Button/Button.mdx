---
menu: Components
name: Button
---

import { Playground, PropsTable } from 'docz';
import Button from './Button';

# Email Button

```js
import Button from '@clarityhub/unity-email/lib/components/Button';
```

## Prop Types

<PropsTable of={Button} />
