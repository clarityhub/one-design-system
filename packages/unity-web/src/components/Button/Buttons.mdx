---
menu: Components
name: Buttons
---

import { Playground, PropsTable } from 'docz';
import DeleteIcon from '@material-ui/icons/Delete';
import CheckIcon from '@material-ui/icons/CheckCircle';
import Button from './index';

# Buttons

Buttons are used for triggering various actions.

```js
import Button from '@clarityhub/unity-web/lib/components/Button';
```

## Button Types

There are 4 main button types: default, white, primary, and danger.

* **Default buttons** should be used for actions that will not cause any changes to happen. For example, a Cancel button may close a modal, but it will not cause any changes to happen.
* **White buttons** are used when you have a non-white background color and you need to use a default button that may have variants. A white button should normally be used the same way a default button would be used.
* **Primary buttons** are for the primary action you would like a user to take. These buttons should not have any destructive consequence to clicking them.
* **Danger buttons** are for actions that would cause destructive consequences to occur; like deleting an item or unscheduling a campaign.

<Playground>
    <div style={{ backgroundColor: '#dde' }}>
        <Button>Default</Button>
        {' '}
        <Button type="white">White</Button>
        {' '}
        <Button type="primary">Primary</Button>
        {' '}
        <Button type="danger">Danger</Button>
    </div>
</Playground>

## As a Link

If you need a React Router Link, use [LinkButton](./packages-unity-web-src-components-link-linkbutton).

Otherwise, you can have buttons become anchor tags.

<Playground>
    <Button component="a" href="#">Cancel</Button>
    {' '}
    <Button component="a" href="#" type="white">Cancel</Button>
    {' '}
    <Button component="a" href="#" type="primary">Create</Button>
    {' '}
    <Button component="a" href="#" type="danger">Delete</Button>
</Playground>

## States

<Playground>
    <Button className="focused">Focused</Button>
    {' '}
    <Button className="focused" type="white">Focused</Button>
    {' '}
    <Button className="focused" type="primary">Focused</Button>
    {' '}
    <Button className="focused" type="danger">Focused</Button>
    <br/>
    <br/>
    <Button className="active">Active</Button>
    {' '}
    <Button className="active" type="white">Active</Button>
    {' '}
    <Button className="active" type="primary">Active</Button>
    {' '}
    <Button className="active" type="danger">Active</Button>
    <br/>
    <br/>
    <Button className="disabled">Disabled</Button>
    {' '}
    <Button className="disabled" type="white">Disabled</Button>
    {' '}
    <Button className="disabled" type="primary">Disabled</Button>
    {' '}
    <Button className="disabled" type="danger">Disabled</Button>
</Playground>

## Loading

The `loading` property should be used with the `disabled` property.

<Playground>
    <Button size="small" loading disabled>Text that is pretty long</Button>
    {' '}
    <Button size="small" loading disabled type="primary">Text that is pretty long</Button>
    {' '}
    <Button size="small" loading disabled type="danger">Text that is pretty long</Button>
    <br/>
    <br/>
    <Button loading disabled>Text that is pretty long</Button>
    {' '}
    <Button loading disabled type="primary">Text that is pretty long</Button>
    {' '}
    <Button loading disabled type="danger">Text that is pretty long</Button>
    <br/>
    <br/>
    <Button size="large" loading disabled>Text that is pretty long</Button>
    {' '}
    <Button size="large" loading disabled type="primary">Text that is pretty long</Button>
    {' '}
    <Button size="large" loading disabled type="danger">Text that is pretty long</Button>
    <br/>
    <br/>
    <Button size="small" outline loading disabled>Text that is pretty long</Button>
    {' '}
    <Button size="small" outline loading disabled type="primary">Text that is pretty long</Button>
    {' '}
    <Button size="small" outline loading disabled type="danger">Text that is pretty long</Button>
    <br/>
    <br/>
    <Button outline loading disabled>Text that is pretty long</Button>
    {' '}
    <Button outline loading disabled type="primary">Text that is pretty long</Button>
    {' '}
    <Button outline loading disabled type="danger">Text that is pretty long</Button>
    <br/>
    <br/>
    <Button size="large" outline loading disabled>Text that is pretty long</Button>
    {' '}
    <Button size="large" outline loading disabled type="primary">Text that is pretty long</Button>
    {' '}
    <Button size="large" outline loading disabled type="danger">Text that is pretty long</Button>
</Playground>

## Sizes

You can change the size of buttons for use in different contexts.

* Small buttons may be used in table views
* The default size of buttons should be used in most cases
* Large buttons can be used for calls to action

<Playground>
    <Button size="small">Small</Button>
    {' '}
    <Button size="small" type="primary">Small</Button>
    {' '}
    <Button size="small" type="danger">Small</Button>
    <br/>
    <br/>
    <Button size="default">Default</Button>
    {' '}
    <Button size="default" type="primary">Default</Button>
    {' '}
    <Button size="default" type="danger">Default</Button>
    <br/>
    <br/>
    <Button size="large">Large</Button>
    {' '}
    <Button size="large" type="primary">Large</Button>
    {' '}
    <Button size="large" type="danger">Large</Button>
</Playground>

## Icon Buttons

Specifying the icon prop will space out the icon appropriately when used with text.

<Playground>
    <Button icon={DeleteIcon}>With Icon</Button>
    {' '}
    <Button icon={<DeleteIcon />} type="primary">With Icon</Button>
    {' '}
    <Button icon={<DeleteIcon />} type="danger">With Icon</Button>
    {' '}
    <Button size="large" icon={<DeleteIcon />}>With Icon</Button>
</Playground>

## Other Variants

There are outline and text variants that can be used in different scenarios.

* Outline buttons should only be used for secondary actions
* Text buttons can be used to de-emphasis an action

<Playground>
    <div style={{ backgroundColor: '#dde' }}>
        <Button outline>Outline</Button>
        {' '}
        <Button outline type="white">White</Button>
        {' '}
        <Button outline type="primary">Outline</Button>
        {' '}
        <Button outline type="danger">Outline</Button>
        <br/>
        <br/>
        <Button disabled outline>Outline</Button>
        {' '}
        <Button disabled outline type="white">White</Button>
        {' '}
        <Button disabled outline type="primary">Outline</Button>
        {' '}
        <Button disabled outline type="danger">Outline</Button>
        <br/>
        <br/>
        <Button text>text</Button>
        {' '}
        <Button text type="white">text</Button>
        {' '}
        <Button text type="primary">text</Button>
        {' '}
        <Button text type="danger">text</Button>
        <br/>
        <br/>
        <Button disabled text>text</Button>
        {' '}
        <Button disabled text type="white">text</Button>
        {' '}
        <Button disabled text type="primary">text</Button>
        {' '}
        <Button disabled text type="danger">text</Button>
        <br/>
        <br/>
        <Button block>Button Block</Button>
        {' '}
        <Button block type="primary">Button Block</Button>
        {' '}
        <Button block type="danger">Button Block</Button>
    </div>
</Playground>

## Prop Types

The following prop types are available for Buttons.

<PropsTable of={Button} />

