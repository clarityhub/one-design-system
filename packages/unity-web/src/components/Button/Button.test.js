import React, { Fragment } from 'react';
import renderer from 'react-test-renderer';

import Button from './Button';

describe('<Button />', () => {
	it('renders solid buttons', () => {
		const node = (
			<Fragment>
				<Button>Default</Button>
				{' '}
				<Button type="white">White</Button>
				{' '}
				<Button type="primary">Primary</Button>
				{' '}
				<Button type="danger">Danger</Button>
			</Fragment>
		);
		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
    
	it('renders outline buttons', () => {
		const node = (
			<Fragment>
				<Button outline>Default</Button>
				{' '}
				<Button outline type="white">White</Button>
				{' '}
				<Button outline type="primary">Primary</Button>
				{' '}
				<Button outline type="danger">Danger</Button>
			</Fragment>
		);
		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
    });
    
    it('renders link buttons', () => {
		const node = (
			<Fragment>
				<Button component="a" href="#">Link</Button>
			</Fragment>
		);
		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
});