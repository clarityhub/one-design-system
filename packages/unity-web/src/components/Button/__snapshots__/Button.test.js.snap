// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders link buttons 1`] = `
<a
  className="-default css-1t30yu8-ButtonWrapper-Button-Button entsyi63"
  href="#"
  type="button"
>
  Link
</a>
`;

exports[`<Button /> renders outline buttons 1`] = `
Array [
  <button
    className="-default -outline css-1qfeqqs-Button-Button entsyi60"
    type="button"
  >
    Default
  </button>,
  " ",
  <button
    className="-white -outline css-1qfeqqs-Button-Button entsyi60"
    type="button"
  >
    White
  </button>,
  " ",
  <button
    className="-primary -outline css-1qfeqqs-Button-Button entsyi60"
    type="button"
  >
    Primary
  </button>,
  " ",
  <button
    className="-danger -outline css-1qfeqqs-Button-Button entsyi60"
    type="button"
  >
    Danger
  </button>,
]
`;

exports[`<Button /> renders solid buttons 1`] = `
Array [
  <button
    className="-default css-1qfeqqs-Button-Button entsyi60"
    type="button"
  >
    Default
  </button>,
  " ",
  <button
    className="-white css-1qfeqqs-Button-Button entsyi60"
    type="button"
  >
    White
  </button>,
  " ",
  <button
    className="-primary css-1qfeqqs-Button-Button entsyi60"
    type="button"
  >
    Primary
  </button>,
  " ",
  <button
    className="-danger css-1qfeqqs-Button-Button entsyi60"
    type="button"
  >
    Danger
  </button>,
]
`;
